File: challenge1.py
# Problem Challenge 1: Permutation in a String (hard)

# Input: String="oidbcaf", Pattern="abc"
# Output: true
# Explanation: The string contains "bca" which is a permutation of the given pattern.

def challenge(s, pattern) -> bool:
    pattern_dict = {}
    matchCount = 0
    window_start = 0
    for v in pattern:
        if v not in pattern_dict:
            pattern_dict[v] = 0
        pattern_dict[v] += 1

    for window_end in range(len(s)):
        right_char = s[window_end]
        if right_char in pattern_dict:
            pattern_dict[right_char] -= 1
            if pattern_dict[right_char] == 0:
                matchCount += 1
        
        if matchCount == len(pattern_dict):
            return True
        
        # shrinking window
        if window_end >= len(pattern) - 1:
            left_char = s[window_start]
            window_start += 1
            if left_char in pattern_dict:
                if pattern_dict[left_char] == 0:
                    matchCount -= 1  # Before putting the character back, decrement the matched count
                pattern_dict[left_char] += 1  # Put the character back
                
    return False


print(challenge("ppqp", "pq"))
print(challenge("abbcabc", "abcd"))

File: challenge1.js
//Problem Challenge 1: Permutation in a String (hard)

// Input: String="oidbcaf", Pattern="abc"
// Output: true
// Explanation: The string contains "bca" which is a permutation of the given pattern.

// { a: 1, b: 1, c: 1}

function permutation(s, pattern) {
    let windowStart = 0,
        matchCount = 0;
    const patternMap = {};

    for (v of pattern) {
        if (!(v in patternMap)) {
            patternMap[v] = 0
        }
        patternMap[v]++;
    }

    for (let windowEnd=0; windowEnd < s.length; windowEnd++) {
        let rightChar = s[windowEnd];
        if (rightChar in patternMap) {
            patternMap[rightChar]--;
            if (patternMap[rightChar] === 0) {
                matchCount++;
            }
        }

        if (matchCount === Object.keys(patternMap).length) {
            return true;
        }

        // shrink window

        if (windowEnd + 1 >= pattern.length) {
            let leftChar = s[windowStart];
            if (leftChar in patternMap) {
                if (patternMap[leftChar] === 0) {
                    matchCount--;
                }
                patternMap[leftChar]++;
            }
            windowStart++;
        }

    }
    return false;
}

console.log(permutation("oidbcaf", "abc"))
console.log(permutation("odicf", "dc"))